File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/usecase/usecases/PublishCoupon.java
package com.mycompany.myshop.testkit.dsl.core.usecase.usecases;

import com.mycompany.myshop.testkit.driver.port.dtos.PublishCouponRequest;
import com.mycompany.myshop.testkit.driver.port.MyShopDriver;
import com.mycompany.myshop.testkit.dsl.core.usecase.usecases.base.BaseMyShopUseCase;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseResult;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseContext;
import com.mycompany.myshop.testkit.dsl.core.shared.VoidVerification;

public class PublishCoupon extends BaseMyShopUseCase<Void, VoidVerification> {
    private String couponCodeParamAlias;
    private String discountRate;
    private String validFrom;
    private String validTo;
    private String usageLimit;

    public PublishCoupon(MyShopDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    public PublishCoupon couponCode(String couponCodeParamAlias) {
        this.couponCodeParamAlias = couponCodeParamAlias;
        return this;
    }

    public PublishCoupon discountRate(String discountRate) {
        this.discountRate = discountRate;
        return this;
    }

    public PublishCoupon validFrom(String validFrom) {
        this.validFrom = validFrom;
        return this;
    }

    public PublishCoupon validTo(String validTo) {
        this.validTo = validTo;
        return this;
    }

    public PublishCoupon usageLimit(String usageLimit) {
        this.usageLimit = usageLimit;
        return this;
    }

    @Override
    public UseCaseResult<Void, VoidVerification> execute() {
        var couponCode = context.getParamValue(couponCodeParamAlias);

        var request = PublishCouponRequest.builder()
                .code(couponCode)
                .discountRate(discountRate)
                .validFrom(validFrom)
                .validTo(validTo)
                .usageLimit(usageLimit)
                .build();

        var result = driver.publishCoupon(request);
        return new UseCaseResult<>(result, context, VoidVerification::new);
    }
}

File: system-test/dotnet/Dsl.Core/UseCase/UseCases/PublishCoupon.cs
using Driver.Port;
using Dsl.Core.UseCase.UseCases.Base;
using Driver.Port.Dtos;
using Driver.Port.Dtos.Error;
using Dsl.Core.Shared;
using Common;

namespace Dsl.Core.UseCase.UseCases;

public class PublishCoupon : BaseMyShopUseCase<VoidValue, VoidVerification>
{
    private string? _couponCodeParamAlias;
    private string? _discountRate;
    private string? _validFrom;
    private string? _validTo;
    private string? _usageLimit;

    public PublishCoupon(IMyShopDriver driver, UseCaseContext context)
        : base(driver, context)
    {
    }

    public PublishCoupon CouponCode(string? couponCodeParamAlias)
    {
        _couponCodeParamAlias = couponCodeParamAlias;
        return this;
    }

    public PublishCoupon DiscountRate(decimal discountRate)
    {
        _discountRate = discountRate.ToString();
        return this;
    }

    public PublishCoupon DiscountRate(string? discountRate)
    {
        _discountRate = discountRate;
        return this;
    }

    public PublishCoupon ValidFrom(string? validFrom)
    {
        _validFrom = validFrom;
        return this;
    }

    public PublishCoupon ValidTo(string? validTo)
    {
        _validTo = validTo;
        return this;
    }

    public PublishCoupon UsageLimit(string? usageLimit)
    {
        _usageLimit = usageLimit;
        return this;
    }

    public override async Task<MyShopUseCaseResult<VoidValue, VoidVerification>> Execute()
    {
        var couponCode = _context.GetParamValue(_couponCodeParamAlias);

        var request = new PublishCouponRequest
        {
            Code = couponCode,
            DiscountRate = string.IsNullOrWhiteSpace(_discountRate) ? null : _discountRate,
            ValidFrom = string.IsNullOrWhiteSpace(_validFrom) ? null : _validFrom,
            ValidTo = string.IsNullOrWhiteSpace(_validTo) ? null : _validTo,
            UsageLimit = string.IsNullOrWhiteSpace(_usageLimit) ? null : _usageLimit
        };

        var result = await _driver.PublishCouponAsync(request);

        return new MyShopUseCaseResult<VoidValue, VoidVerification>(
            result,
            _context,
            (response, ctx) => new VoidVerification(response, ctx));
    }
}
